include/pmacc/memory/boxes/PitchedBox.hpp

```diff
@@ -23,154 +23,195 @@
-    class PitchedBox<TYPE, DIM2> : public detail::PitchedBoxCommon<TYPE, DIM2>
-    {
-        using Base = detail::PitchedBoxCommon<TYPE, DIM2>;
+        /*Object must init by copy a valid instance*/
+        PitchedBox() = default;
 
-    public:
-        using ReducedType = PitchedBox<TYPE, DIM1>;
+        /** Constructor
+         *
+         * @tparam T_MemoryIdxType Index type
+         * @param pointer pointer to the memory
+         * @param extent extent of the memory in elements
+         * @param pitch line pitch in bytes
+         */
+        template<typename T_MemoryIdxType>
+        HDINLINE PitchedBox(T_Type* pointer, math::Vector<T_MemoryIdxType, T_dim> const& extent, size_t const pitch)
+            : m_ptr(pointer)
+        {
+            m_pitch.x() = pitch;
+            for(uint32_t d = 1; d < T_dim - 1; ++d)
+                m_pitch[d] = m_pitch[d - 1u] * static_cast<size_t>(extent[d]);
+        }
 
-        /*Object must init by copy a valid instance*/
-        HDINLINE PitchedBox() = default;
+        PitchedBox(PitchedBox const&) = default;
 
-        HDINLINE PitchedBox(TYPE* pointer, size_t pitch) : Base{pointer}, pitch(pitch)
+        /** get value at the given index
+         *
+         * @tparam T_MemoryIdxType Index type
+         * @param idx n-dimensional offset, relative to the origin pointer
+         * @return reference to the value
+         * @{
+         */
+        template<typename T_MemoryIdxType>
+        HDINLINE T_Type const& operator[](math::Vector<T_MemoryIdxType, T_dim> const& idx) const
         {
+            return *ptr(idx);
         }
 
-        HDINLINE PitchedBox(TYPE* pointer, const DataSpace<DIM2>& /*memSize*/, const size_t pitch)
-            : Base{pointer}
-            , pitch(pitch)
+        template<typename T_MemoryIdxType>
+        HDINLINE T_Type& operator[](math::Vector<T_MemoryIdxType, T_dim> const& idx)
         {
+            return *const_cast<T_Type*>(ptr(idx));
         }
 
-        HDINLINE PitchedBox(PitchedBox const&) = default;
+        /** }@ */
 
-        HDINLINE ReducedType operator[](const int idx) const
+    protected:
+        /** get the pointer of the value relative to the origin pointer m_ptr
+         *
+         * @tparam T_MemoryIdxType Index type
+         * @param idx n-dimensional offset
+         * @return pointer to value
+         */
+        template<typename T_MemoryIdxType>
+        HDINLINE T_Type const* ptr(math::Vector<T_MemoryIdxType, T_dim> const& idx) const
         {
-            return ReducedType((TYPE*) ((char*) this->fixedPointer + idx * pitch));
+            /** offset in bytes
+             *
+             * We calculate the complete offset in bytes even if it would be possible to change the x-dimension with
+             * the native types pointer, this is reducing the register footprint.
+             */
+            size_t offset = sizeof(T_Type) * idx.x();
+            for(uint32_t d = 1u; d < T_dim; ++d)
```

Comment:
same as below/above:
This only iterates over `T_dim -1` values. Is this intended?

Reply:
Yes the pitch extents (in bytes) is one dimension less compared to the index dimension.